geode-core/src/main/java/org/apache/geode/internal/cache/TXCommitMessage.java

```diff
@@ -1384,7 +1384,11 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
         for (int i = 0; i < size; i++) {
           FarSideEntryOp entryOp = new FarSideEntryOp();
           // shadowkey is not being sent to clients
-          entryOp.fromData(in, largeModCount, !this.msg.getDM().isLoner());
+          LocalRegion region = getRegionByPath(msg.getDM(),
+              parentRegionPath != null ? parentRegionPath : regionPath);
+          boolean readShadowKey =
+              (region != null ? (region.getPoolName() == null) : !msg.getDM().isLoner());
```

Comment:
These two lines with their ternaries are so dense with logic, that I wonder if it might be worthwhile to split this out into a private method and break this out more into variables and maybe even a if/else statement instead of the ternary to make this easier to follow.